### PR TITLE
[SYCL][ESIMD] Add ESIMD PrefixSum kernel tests.

### PR DESCRIPTION
**Pull request by kychendev**

### CODE REVIEW

File: sycl/test/esimd/on-device/PrefixSum.cpp [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+#elif TUPLE_SZ == 4
+#define GATHER_SCATTER_MASK ESIMD_ABGR_ENABLE
+#endif
+
+#define LOG_ENTRIES 8
+#define PREFIX_ENTRIES (1 << LOG_ENTRIES)
+#define PREFIX_ENTRIES_LOW 32
+#define ENTRIES_THRESHOLD 2048
+#define MIN_NUM_THREADS                                                        \
+  1 // minimum number of threads to launch a kernel (power of 2)
```
Comment (s-kanaev) [inline sycl/test/esimd/on-device/PrefixSum.cpp:36]: This is no good. The macro will expand into `1 // and the comment` which may lead to unexpected and unwanted consequences.

File: sycl/test/esimd/on-device/PrefixSum.cpp [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+#define MAX_TS_WIDTH 1024
+#define TUPLE_SZ 1 // kernel can only handle TUPLE_SZ can be 1, 2, 4
+
+#if TUPLE_SZ == 1
+#define GATHER_SCATTER_MASK ESIMD_R_ENABLE
+#elif TUPLE_SZ == 2
+#define GATHER_SCATTER_MASK ESIMD_GR_ENABLE
+#elif TUPLE_SZ == 4
+#define GATHER_SCATTER_MASK ESIMD_ABGR_ENABLE
+#endif
```
Comment (s-kanaev) [inline sycl/test/esimd/on-device/PrefixSum.cpp:29]: Should we also have `RUN`'s for the test with different `TUPLE_SZ`?

File: sycl/test/esimd/on-device/PrefixSum.cpp [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+int main(int argc, char *argv[]) {
+
+  unsigned int *pInputs;
+#if 0
+    if (argc < 2) {
+        std::cout << "Usage: prefix [N]. N is 2^N entries x TUPLE_SZ" << std::endl;
+        exit(1);
+    }
+    unsigned log2_element = atoi(argv[1]);
+#endif
```
Comment (s-kanaev) [inline sycl/test/esimd/on-device/PrefixSum.cpp:392]: Please remove these as it makes hard to read the code.
Though, passing arguments to test binary is nice way to improve its debuggability.

**Review by s-kanaev — COMMENTED** [review]
I have only looked into `PrefixSum.cpp` test.
If I understood correctly the test is still under development.
If so, please move it to draft or mark as `[WIP]` in caption.

File: sycl/test/esimd/on-device/PrefixSum.cpp [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+  compute_prefixsum_up(pExpectOutputs, size, 1, PREFIX_ENTRIES);
+#endif
+#endif
+
+  hierarchical_prefix(q, pInputs, 1, PREFIX_ENTRIES, size, PREFIX_ENTRIES);
+  q.wait();
+
+  bool pass = memcmp(pInputs, pExpectOutputs,
+                     size * TUPLE_SZ * sizeof(unsigned int)) == 0;
+#if 0
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/PrefixSum.cpp:452]: Maybe remove dead code?

File: sycl/test/esimd/on-device/PrefixSum.cpp [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+  // allocate & compute expected result
+  unsigned int *pExpectOutputs = static_cast<unsigned int *>(
+      malloc(size * TUPLE_SZ * sizeof(unsigned int)));
+
+#if 1
+  compute_prefixsum(pInputs, pExpectOutputs, size);
+#else // validate intermediate steps
+  memcpy(pExpectOutputs, pInputs, size * TUPLE_SZ * sizeof(unsigned));
+#if 0 // 1 << 26 entries
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/PrefixSum.cpp:420]: Maybe remove dead code?

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+  pInputs = static_cast<unsigned int *>(
+      malloc_shared(size * TUPLE_SZ * sizeof(unsigned int), dev, ctxt));
+  for (unsigned int i = 0; i < size * TUPLE_SZ; ++i) {
+    pInputs[i] = rand() % 128;
+  }
+
+  // allocate & compute expected result
+  unsigned int *pExpectOutputs = static_cast<unsigned int *>(
+      malloc(size * TUPLE_SZ * sizeof(unsigned int)));
+  // compute_prefixsum(pInputs, pExpectOutputs, size);
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum1.cpp:170]: please remove

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        local_sum[j] += input[(k + i) * TUPLE_SZ + j];
+      }
+    }
+    // store local_sum in the last entry
+    memcpy(&prefixSum[(k + PREFIX_ENTRIES - 1) * TUPLE_SZ], local_sum,
+           TUPLE_SZ * sizeof(unsigned));
+  }
+}
+
+void compute_prefixsum(unsigned int input[], unsigned int prefixSum[],
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum1.cpp:60]: This code is unused?

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+void cmk_sum_tuple_count(unsigned int *buf, unsigned int h_pos) {
+  // h_pos indicates which 256-element chunk the kernel is processing
+
+  // each thread handles PREFIX_ENTRIES entries. Each entry has 4 bins
+  unsigned int offset = h_pos * PREFIX_ENTRIES * TUPLE_SZ;
+
+  simd<unsigned, 32 * TUPLE_SZ> S, T;
+#pragma unroll
+  for (int i = 0; i < TUPLE_SZ; i++) {
+    // select() is not assignable. should fix this in the future
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum1.cpp:87]: Add a `TODO` mark please.

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    simd<unsigned, 32> tmp = block_load<unsigned, 32>(buf + offset + i * 32);
+    S.select<32, 1>(i * 32) = tmp;
+  }
+
+#pragma unroll
+  for (int i = 1; i < PREFIX_ENTRIES / 32; i++) {
+#pragma unroll
+    for (int j = 0; j < TUPLE_SZ; j++) {
+      // select() is not assignable
+      // T.select<32, 1>(j * 32) = block_load<unsigned, 32>(buf + offset + i *
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum1.cpp:99]: `TODO` mark

File: sycl/test/esimd/on-device/Prefix_Local_sum2.cpp [hunk 10]
```diff
@@ -0,0 +1,197 @@
+//==-------- Prefix_Local_sum2.cpp  - DPC++ ESIMD on-device test
+//------------==//
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum2.cpp:2]: minor: fix file header.

File: sycl/test/esimd/on-device/Prefix_Local_sum2.cpp [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        local_sum[j] += input[(k + i) * TUPLE_SZ + j];
+      }
+    }
+    // store local_sum in the last entry
+    memcpy(&prefixSum[(k + PREFIX_ENTRIES - 1) * TUPLE_SZ], local_sum,
+           TUPLE_SZ * sizeof(unsigned));
+  }
+}
+
+void compute_prefixsum(unsigned int input[], unsigned int prefixSum[],
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum2.cpp:60]: unused?

File: sycl/test/esimd/on-device/Prefix_Local_sum3.cpp [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+    for (int j = 0; j < TUPLE_SZ; j++) {
+      local_sum[j] += prefixSum[((i + 1) * elem_stride - 1) * TUPLE_SZ + j];
+    }
+    // update every elem_stride entry
+    memcpy(&prefixSum[((i + 1) * elem_stride - 1) * TUPLE_SZ], local_sum,
+           TUPLE_SZ * sizeof(unsigned));
+  }
+}
+
+void compute_prefixsum(unsigned int input[], unsigned int prefixSum[],
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/on-device/Prefix_Local_sum3.cpp:80]: unused?

**Review by DenisBakhvalov — CHANGES_REQUESTED** [review]
I would suggest adding a short description of what those tests do and how they differ from each other.

File: sycl/test/esimd/on-device/PrefixSum.cpp (hunk 1 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/PrefixSum.cpp]: Fixed.

File: sycl/test/esimd/on-device/PrefixSum.cpp (hunk 2 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/PrefixSum.cpp:30]: This is a demo case for implementing PrefixSum in ESIMD, with configurable parameters for tuning purpose. Some different TUPLE_SIZE are exercised in other variants of this kernel.

File: sycl/test/esimd/on-device/PrefixSum.cpp (hunk 3 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/PrefixSum.cpp]: Changed to use input argument.

File: sycl/test/esimd/on-device/PrefixSum.cpp (hunk 4 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/PrefixSum.cpp]: Cleaned up.

File: sycl/test/esimd/on-device/PrefixSum.cpp (hunk 5 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/PrefixSum.cpp]: Cleaned up.

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp (hunk 6 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum1.cpp]: Cleaned up.

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp (hunk 7 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum1.cpp]: Cleaned up.

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp (hunk 8 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum1.cpp]: Fixed.

File: sycl/test/esimd/on-device/Prefix_Local_sum1.cpp (hunk 9 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum1.cpp]: Fixed.

File: sycl/test/esimd/on-device/Prefix_Local_sum2.cpp (hunk 10 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum2.cpp]: Fixed.

File: sycl/test/esimd/on-device/Prefix_Local_sum2.cpp (hunk 11 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum2.cpp]: Cleaned up.

File: sycl/test/esimd/on-device/Prefix_Local_sum3.cpp (hunk 12 above)
Comment (kychendev) [reply sycl/test/esimd/on-device/Prefix_Local_sum3.cpp]: Cleaned up.

**Comment by DenisBakhvalov** [comment]
@kychendev ESIMD tests were moved to llvm-test-suite (https://github.com/intel/llvm/pull/2611).
I think you need to open a PR in https://github.com/intel/llvm-test-suite

**Comment by kychendev** [comment]
@DenisBakhvalov, thanks for clarification. I have created a new PR for llvm-test-suite. Closing this PR.

**Pull request closed without merging**